test/downloadutils.mjs

```diff
@@ -25,89 +24,74 @@ function rewriteWebArchiveUrl(url) {
-          callback("Too many redirects");
+  return new Promise((resolve, reject) => {
+    protocol
+      .get(url, async function (response) {
+        if ([301, 302, 307, 308].includes(response.statusCode)) {
+          if (redirects > 10) {
+            reject(new Error("Too many redirects"));
+            return;
+          }
+          const redirectTo = urlResolve(url, response.headers.location);
+          try {
+            await downloadFile(file, redirectTo, (redirects || 0) + 1);
+            resolve();
+          } catch (ex) {
+            reject(ex);
+          }
+          return;
         }
-        var redirectTo = response.headers.location;
-        redirectTo = urlResolve(url, redirectTo);
-        downloadFile(file, redirectTo, callback, (redirects || 0) + 1);
-        return;
-      }
 
-      if (response.statusCode !== 200) {
-        callback("HTTP " + response.statusCode);
-        return;
-      }
-      var stream = fs.createWriteStream(file);
-      stream.on("error", function (err) {
-        callback(err);
-      });
-      response.pipe(stream);
-      stream.on("finish", function () {
-        stream.end();
-        callback();
-      });
-    })
-    .on("error", function (err) {
-      callback(err);
-    });
-}
+        if (response.statusCode !== 200) {
+          reject(new Error(`HTTP ${response.statusCode}`));
+          return;
+        }
 
-function downloadManifestFiles(manifest, callback) {
-  function downloadNext() {
-    if (i >= links.length) {
-      callback();
-      return;
-    }
-    var file = links[i].file;
-    var url = links[i].url;
-    console.log("Downloading " + url + " to " + file + "...");
-    downloadFile(file, url, function (err) {
-      if (err) {
-        console.error("Error during downloading of " + url + ": " + err);
-        fs.writeFileSync(file, ""); // making it empty file
-        fs.writeFileSync(file + ".error", err);
-      }
-      i++;
-      downloadNext();
-    });
-  }
+        const stream = fs.createWriteStream(file);
+        stream.on("error", error => reject(error));
+        stream.on("finish", () => {
+          stream.end();
+          resolve();
+        });
+        response.pipe(stream);
+      })
+      .on("error", error => reject(error));
+  });
+}
 
-  var links = manifest
-    .filter(function (item) {
-      return item.link && !fs.existsSync(item.file);
-    })
-    .map(function (item) {
-      var file = item.file;
-      var linkfile = file + ".link";
-      var url = fs.readFileSync(linkfile).toString();
+async function downloadManifestFiles(manifest) {
+  const links = manifest
+    .filter(item => item.link && !fs.existsSync(item.file))
+    .map(item => {
+      let url = fs.readFileSync(`${item.file}.link`).toString();
       url = url.replace(/\s+$/, "");
-      return { file, url };
+      return { file: item.file, url };
     });
 
-  var i = 0;
-  downloadNext();
+  for (const link of links) {
+    const { file, url } = link;
```

Comment:
Nit: If you want to, these lines could be combined as follows.
```suggestion
  for (const { file, url } of links) {
```

Reply:
Good point; will do.